src/Bedrock.Framework.Experimental/Protocols/Kafka/KafkaClientConnectionContext.cs

```diff
@@ -0,0 +1,131 @@
+            _ = WriteRequestsAsync(connectionContext, cancellationToken);
+            _ = ReadResponsesAsync(connectionContext, cancellationToken);
+
+            return Task.CompletedTask;
+        }
+
+        /// <inheritdoc/>
+        public async Task<TResponseSchema> SendAsync<TRequestSchema, TResponseSchema>(
+            TRequestSchema request,
+            IKafkaMessageDescriptor messageDescriptor,
+            CancellationToken cancellationToken = default)
+        {
+            var tcs = new TaskCompletionSource<object>();
+
+            _requestChannel.Writer.TryWrite((request, messageDescriptor, tcs));
+
+            return (TResponseSchema)await tcs.Task;
+        }
+
+        private async Task WriteRequestsAsync(ConnectionContext connectionContext, CancellationToken cancellationToken = default)
+        {
+            var writer = connectionContext.CreateWriter();
+
+            while (await _requestChannel.Reader.WaitToReadAsync(cancellationToken).ConfigureAwait(false))
+            {
+                if (_requestChannel.Reader.TryRead(out (object request, IKafkaMessageDescriptor messageDescriptor, TaskCompletionSource<object> tcs) request))
+                {
+                    var bufferWriter = new MemoryBufferWriter<byte>();
+                    var correlationId = ++_correlationId;
+                    var requestWriter = request.messageDescriptor.GetRequestWriter();
+
+                    _requestQueue.Enqueue((correlationId, request.messageDescriptor, request.tcs));
+
+                    requestWriter.WriteMessage(request, bufferWriter);
+
+                    using var kafkaRequest = new KafkaRequest(bufferWriter, request.messageDescriptor.ApiKey, request.messageDescriptor.ApiVersion, correlationId);
+
+                    await writer.WriteAsync(_requestWriter, kafkaRequest, cancellationToken);
+                }
+            }
+        }
+
+        private async Task ReadResponsesAsync(ConnectionContext connectionContext, CancellationToken cancellationToken = default)
+        {
+            var reader = connectionContext.CreateReader();
+
+            while (true)
+            {
+                var result = await reader.ReadAsync(_responseReader, cancellationToken);
+
+                if (result.IsCompleted)
+                    break;
+
+                using var kafkaResponse = result.Message;
+
+                var hasRequestInFlight = _requestQueue.TryDequeue(
+                    out (int correlationId, IKafkaMessageDescriptor messageDescriptor, TaskCompletionSource<object> tcs) inflightRequestContext);
+
+                Debug.Assert(hasRequestInFlight, "Protocol Error: received unexpected message from server.");
+
+                try
+                {
+                    if (kafkaResponse.CorrelationId != inflightRequestContext.correlationId)
+                    {
+                        inflightRequestContext.tcs.SetException(new Exception("Protocol Exception: correlation ID mis-match."));
```

Comment:
Do you want to terminate the connection at this point, if you are getting mismatches then likely the connection is broken for the rest of time.

Therefore I would have if(!hasRequestInFlight || kafkaResponse.CorrelationId != inflightRequest.correlationId)
then you could set the exception (I would possibly call a "cleanup" or similar method where you complete the channel so you allow no more inflight requests, error out all the outstanding requests and close the connection) ?

Depends on how you want to deal with this situation